x-pack/plugins/apm/server/lib/services/get_services/get_services_items.ts

```diff
@@ -3,88 +3,80 @@
  * or more contributor license agreements. Licensed under the Elastic License;
  * you may not use this file except in compliance with the Elastic License.
  */
-
-import { mergeProjection } from '../../../../common/projections/util/merge_projection';
-import {
-  PROCESSOR_EVENT,
-  AGENT_NAME,
-  SERVICE_ENVIRONMENT,
-  TRANSACTION_DURATION,
-} from '../../../../common/elasticsearch_fieldnames';
+import { uniq } from 'lodash';
+import { arrayUnionToCallable } from '../../../../common/utils/array_union_to_callable';
 import { PromiseReturnType } from '../../../../typings/common';
 import {
   Setup,
   SetupTimeRange,
   SetupUIFilters,
 } from '../../helpers/setup_request';
 import { getServicesProjection } from '../../../../common/projections/services';
+import {
+  getTransactionDurationAvg,
+  getAgentName,
+  getTransactionRate,
+  getErrorRate,
+  getEnvironments,
+} from './get_metrics';
 
 export type ServiceListAPIResponse = PromiseReturnType<typeof getServicesItems>;
-export async function getServicesItems(
-  setup: Setup & SetupTimeRange & SetupUIFilters
-) {
-  const { start, end, client } = setup;
-
-  const projection = getServicesProjection({ setup });
-
-  const params = mergeProjection(projection, {
-    body: {
-      size: 0,
-      aggs: {
-        services: {
-          terms: {
-            ...projection.body.aggs.services.terms,
-            size: 500,
-          },
-          aggs: {
-            avg: {
-              avg: { field: TRANSACTION_DURATION },
-            },
-            agents: {
-              terms: { field: AGENT_NAME, size: 1 },
-            },
-            events: {
-              terms: { field: PROCESSOR_EVENT },
-            },
-            environments: {
-              terms: { field: SERVICE_ENVIRONMENT },
-            },
-          },
-        },
-      },
-    },
-  });
-
-  const resp = await client.search(params);
-  const aggs = resp.aggregations;
-
-  const serviceBuckets = aggs?.services.buckets || [];
-
-  const items = serviceBuckets.map((bucket) => {
-    const eventTypes = bucket.events.buckets;
+export type ServicesItemsSetup = Setup & SetupTimeRange & SetupUIFilters;
+export type ServicesItemsProjection = ReturnType<typeof getServicesProjection>;
 
-    const transactions = eventTypes.find((e) => e.key === 'transaction');
-    const totalTransactions = transactions?.doc_count || 0;
+export async function getServicesItems(setup: ServicesItemsSetup) {
+  const params = {
+    projection: getServicesProjection({ setup, noEvents: true }),
```

Comment:
What does `noEvents` mean? Is this to simulate a no-op?

Reply:
not a no-op, but with `noEvents: true` it doesn't make a decision about what events should be queried. It's not great, but the alternatives (having to define indices + processor.event term filters in a lot of other places) are worse IMO. It made me think though, our client and our projections could support `{ apm: { events: ProcessorEvent[] }` as a top-level prop, and prohibit the use of `index`. The client's search function would then unpack `apm.events` to an index and a terms query.

Reply:
> It made me think though, our client and our projections could support `{ apm: { events: ProcessorEvent[] }` as a top-level prop, and prohibit the use of `index`. The client's search function would then unpack `apm.events` to an index and a terms query.

I think it's a good idea to prohibit the use of index and instead use `apm.events` to specify both the indicies to query and the `processor.events` to filter by.
This is similar to what I tried to suggest in https://github.com/elastic/kibana/issues/67397 - although I called it `docTypes` and was probably not super clear. I've renamed it to `apmEvents` now.

Reply:
@sqren I think I understand the need better now 👍 and IMO a top-level prop lends itself better to composition than a second function argument.

Reply:
In addition to the client idea, we should move away from the mega object `setup` has become and I'd like to move away from `mergeProjections` and the overuse of spreading: https://github.com/elastic/kibana/pull/70157